tsdb/store.go

```diff
@@ -1180,40 +1180,57 @@ func (s *Store) sketchesForDatabase(dbName string, getSketches func(*Shard) (est
 // Cardinality is calculated exactly by unioning all shards' bitsets of series
 // IDs. The result of this method cannot be combined with any other results.
 //
-func (s *Store) SeriesCardinality(database string) (int64, error) {
+func (s *Store) SeriesCardinality(ctx context.Context, database string) (int64, error) {
 	s.mu.RLock()
 	shards := s.filterShards(byDatabase(database))
 	s.mu.RUnlock()
 
 	var setMu sync.Mutex
 	others := make([]*SeriesIDSet, 0, len(shards))
 
-	s.walkShards(shards, func(sh *Shard) error {
-		index, err := sh.Index()
-		if err != nil {
-			return err
-		}
+	err := s.walkShards(shards, func(sh *Shard) error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			index, err := sh.Index()
+			if err != nil {
+				return err
+			}
 
-		seriesIDs := index.SeriesIDSet()
-		setMu.Lock()
-		others = append(others, seriesIDs)
-		setMu.Unlock()
+			seriesIDs := index.SeriesIDSet()
+			setMu.Lock()
+			others = append(others, seriesIDs)
+			setMu.Unlock()
 
-		return nil
+			return nil
+		}
 	})
-
+	if err != nil {
+		return 0, err
+	}
 	ss := NewSeriesIDSet()
 	ss.Merge(others...)
-	return int64(ss.Cardinality()), nil
+	select {
```

Comment:
I don't know if this bail out is necessary - I'd assume by this point there's a minimal amount of processing remaining so we might as well just return the cardinality.

Reply:
My timings showed that the Cardinality computation could take a modicum of time, so I dropped in this check.  Agreed, it's a later bail-out that the earlier checks.